consume.go

```diff
@@ -39,49 +39,130 @@ type consumeCmd struct {
+		anchor: r.startAnchor,
+		diff:   r.diff,
 	}
+}
 
-	var (
-		res int64
-		err error
-	)
+func (r positionRange) end() position {
+	return position{
+		anchor: r.endAnchor,
+		diff:   r.diff,
+	}
+}
+
+// position represents a position within the Kafka stream.
+// It holds an anchor (an absolute position specified as a
+// time stamp or an offset) and a relative position with
+// respect to that (specified as an offset delta or a time duration).
+type position struct {
+	anchor anchor
+	diff   anchorDiff
+}
+
+// anchor represents an absolute offset in the position stream.
+type anchor struct {
+	// isTime specifies which anchor field is valid.
+	// If it's true, the anchor is specified as a time
+	// in time; otherwise it's specified as
+	// an offset in offset.
+	isTime bool
+
+	// offset holds the anchor as an absolute offset.
+	// It can be one of sarama.OffsetOldest, sarama.OffsetNewest
+	// or offsetResume to signify a relative starting position.
+	// This field is only significant when isTime is false.
+	offset int64
+
+	// time holds the anchor as a time.
+	// This field is only significant when isTime is true.
+	time time.Time
+}
+
+func (a0 anchor) eq(a1 anchor) bool {
+	if a0.isTime != a1.isTime {
+		return false
+	}
+	if a0.isTime {
+		return a0.time.Equal(a1.time)
+	}
+	return a0.offset == a1.offset
+}
+
+// anchorDiff represents an offset from an anchor position.
+type anchorDiff struct {
+	// diffIsTime specifies which diff field is valid.
+	// If it's true, the difference is specified as a duration
+	// in the duration field; otherwise it's specified as
+	// an offset in offset.
+	isDuration bool
+
+	// offset holds the difference as an offset delta.
+	offset int64
+
+	// time holds the difference as a duration.
+	duration time.Duration
+}
 
-	if o.start == sarama.OffsetNewest || o.start == sarama.OffsetOldest {
-		if res, err = cmd.client.GetOffset(cmd.topic, partition, o.start); err != nil {
+// timeRange holds a time range.
+// This represents the precision specified in a timestamp
+// (for example, when a time is specified as a date,
+// the time range will include the whole of that day).
+// TODO is this exclusive or inclusive?
+type timeRange struct {
+	t0, t1 time.Time
+}
+
+func (r timeRange) add(d time.Duration) timeRange {
+	return timeRange{
+		t0: r.t0.Add(d),
+		t1: r.t1.Add(d),
+	}
+}
+
+type interval struct {
+	start position
+	end   position
+}
+
+func (cmd *consumeCmd) resolveOffset(p position, partition int32) (int64, error) {
+	if p.anchor.isTime || p.diff.isDuration {
+		return 0, fmt.Errorf("time-based positions not yet supported")
+	}
+	var startOffset int64
+	switch p.anchor.offset {
+	case sarama.OffsetNewest, sarama.OffsetOldest:
+		off, err := cmd.client.GetOffset(cmd.topic, partition, p.anchor.offset)
+		if err != nil {
 			return 0, err
 		}
-
-		if o.start == sarama.OffsetNewest {
-			res = res - 1
+		if p.anchor.offset == sarama.OffsetNewest {
+			// TODO add comment explaining this.
```

Comment:
I'm not sure I understand it myself! It's from the original code, and I added the TODO because I don't understand that code. I'll resolve the issue in a later PR.